simulator/task_manager.go

```diff
@@ -109,3 +179,220 @@ func (m *TaskManager) CreateTask(ctx *Context, req *types.CreateTask) soap.HasFa
 
 	return body
 }
+
+type TaskHistoryCollector struct {
+	mo.TaskHistoryCollector
+
+	*HistoryCollector
+}
+
+func (m *TaskManager) createCollector(ctx *Context, req *types.CreateCollectorForTasks) (*TaskHistoryCollector, *soap.Fault) {
+	if len(m.history.collectors) >= int(m.MaxCollector) {
+		return nil, Fault("Too many task collectors to create", new(types.InvalidState))
+	}
+
+	collector := &TaskHistoryCollector{
+		HistoryCollector: newHistoryCollector(ctx, m.history, defaultPageSize),
+	}
+	collector.Filter = req.Filter
+
+	return collector, nil
+}
+
+// taskFilterChildren returns true if a child of self is the task entity.
+func taskFilterChildren(ctx *Context, root types.ManagedObjectReference, task *types.TaskInfo) bool {
+	seen := false
+
+	var match func(types.ManagedObjectReference)
+
+	match = func(child types.ManagedObjectReference) {
+		if child == *task.Entity {
+			seen = true
+			return
+		}
+
+		walk(ctx.Map.Get(child), match)
+	}
+
+	walk(ctx.Map.Get(root), match)
+
+	return seen
+}
+
+// entityMatches returns true if the spec Entity filter matches the task entity.
+func (c *TaskHistoryCollector) entityMatches(ctx *Context, task *types.TaskInfo, spec *types.TaskFilterSpec) bool {
+	e := spec.Entity
+	if e == nil {
+		return true
+	}
+
+	isSelf := *task.Entity == e.Entity
+
+	switch e.Recursion {
+	case types.TaskFilterSpecRecursionOptionSelf:
+		return isSelf
+	case types.TaskFilterSpecRecursionOptionChildren:
+		return taskFilterChildren(ctx, e.Entity, task)
+	case types.TaskFilterSpecRecursionOptionAll:
+		return isSelf || taskFilterChildren(ctx, e.Entity, task)
+	}
+
+	return false
+}
+
+func (c *TaskHistoryCollector) stateMatches(_ *Context, task *types.TaskInfo, spec *types.TaskFilterSpec) bool {
+	if len(spec.State) == 0 {
+		return true
+	}
+
+	for _, state := range spec.State {
+		if task.State == state {
+			return true
+
+		}
+	}
+
+	return false
+}
+
+func (c *TaskHistoryCollector) timeMatches(_ *Context, task *types.TaskInfo, spec *types.TaskFilterSpec) bool {
+	if spec.Time == nil {
+		return true
+	}
+
+	created := task.QueueTime
+
+	if begin := spec.Time.BeginTime; begin != nil {
+		if created.Before(*begin) {
+			return false
+		}
+	}
+
+	if end := spec.Time.EndTime; end != nil {
+		if created.After(*end) {
+			return false
+		}
+	}
+
+	return true
+}
+
+// taskMatches returns true one of the filters matches the task.
+func (c *TaskHistoryCollector) taskMatches(ctx *Context, task *types.TaskInfo) bool {
+	spec := c.Filter.(types.TaskFilterSpec)
+
+	matchers := []func(*Context, *types.TaskInfo, *types.TaskFilterSpec) bool{
+		c.stateMatches,
+		c.timeMatches,
+		c.entityMatches,
+		// TODO: spec.UserName, etc
```

Comment:
Should any of the fields that we do not yet consider be allowed when creating the filter? Should an InvalidState or ServerError error be returned from the simulator to prevent folks from expecting this filter configs to work?

Reply:
Have the same TODO in EventManager.. it probably isn't much more work to implement the matchers, than to figure out which spec fields to error on. Just kept it to the common use-cases for now.